src/CloseProtectionStream.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace Legionth\React\Http;
+
+use Evenement\EventEmitter;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\WritableStreamInterface;
+
+/**
+ * Protects the passed ReadableStream to be closed
+ */
+class CloseProtectionStream extends EventEmitter implements ReadableStreamInterface
+{
+    private $connection;
+    private $closed = false;
+
+    public function __construct(ReadableStreamInterface $input)
+    {
+        $this->input = $input;
+
+        $this->input->on('data', array($this, 'handleData'));
+        $this->input->on('end', array($this, 'handleEnd'));
+        $this->input->on('error', array($this, 'handleError'));
+        $this->input->on('close', array($this, 'close'));
+    }
+
+    public function isReadable()
+    {
+        return $this->input->isReadable();
+    }
+
+    public function pause()
+    {
+        $this->input->pause();
+    }
+
+    public function resume()
+    {
+        $this->input->resume();
+    }
+
+    public function pipe(WritableStreamInterface $dest, array $options = array())
+    {
+        return $this->input->pipe($dest, $options);
+    }
+
+     public function close()
+     {
+         if ($this->closed) {
+             return;
+         }
+
+         $this->closed = true;
+
+         // ???
+         $this->input->pause();
+
+         $this->readable = false;
+
+         $this->emit('end', array($this));
+         $this->emit('close', array($this));
+         $this->removeAllListeners();
```

Comment:
Note that this will only remove subscribes from this output stream.

This instance still listens for events from the input stream, so perhaps we should remove all of our listeners as well (possibly before)? (needs tests)